mm2src/coins/z_coin/storage/walletdb/mod.rs

```diff
@@ -33,27 +46,96 @@ pub struct WalletDbShared {
+        height: block.height as u32,
+        hash,
+        time: block.time,
+        sapling_tree,
+    }))
+}
+
 #[cfg(not(target_arch = "wasm32"))]
 impl<'a> WalletDbShared {
-    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>) -> MmResult<Self, WalletDbError> {
+    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>, rpc: &mut impl ZRpcOps) -> MmResult<Self, WalletDbError> {
         let z_spending_key = match zcoin_builder.z_spending_key {
             Some(ref z_spending_key) => z_spending_key.clone(),
             None => extended_spending_key_from_protocol_info_and_policy(
                 &zcoin_builder.protocol_info,
                 &zcoin_builder.priv_key_policy,
             )
-            .map_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
+            .mm_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
         };
+
+        let current_block_height = rpc
+            .get_block_height()
+            .await
+            .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+
+        let sync_block = match zcoin_builder.z_coin_params.mode.clone() {
+            ZcoinRpcMode::Light { sync_params, .. } => {
+                let sync_height = match sync_params {
+                    Some(params) => match params {
+                        LightClientSyncParams::Date(date) => zcoin_builder
+                            .calculate_starting_height_from_date(date, current_block_height)
+                            .map_err(|err| {
+                                WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(err))
+                            })?,
+                        LightClientSyncParams::Height(height) => height,
+                    },
+                    None => {
+                        let avg_blocktime = zcoin_builder.conf()["avg_blocktime"].as_u64().ok_or_else(|| {
+                            WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(format!(
+                                "avg_blocktime not  specified in {} coin config",
+                                zcoin_builder.ticker()
+                            )))
+                        })?;
+                        let buffer = avg_blocktime * DAY_IN_HOURS;
+
+                        if current_block_height < buffer {
+                            return MmError::err(WalletDbError::ZcoinClientInitError(
+                                ZcoinClientInitError::ZcashDBError(format!(
+                                    "{} current_block_height: {current_block_height} must be greater than buffer: \
+                                        {buffer}",
+                                    zcoin_builder.ticker()
+                                )),
+                            ));
+                        }
+
+                        current_block_height - buffer
+                    },
+                };
+
+                checkpoint_block_from_height(sync_height, rpc).await?
+            },
```

Comment:
This can be simplified as
```rust
ZcoinRpcMode::Light { sync_params, .. } => {
    let sync_height = match sync_params {
        Some(LightClientSyncParams::Date(date)) => zcoin_builder
            .calculate_starting_height_from_date(date, current_block_height)
            .map_err(|err| WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(err)))?,
        Some(LightClientSyncParams::Height(height)) => height,
        None => zcoin_builder
            .calculate_starting_height_from_date(now_sec() - DAY_IN_SECONDS, current_block_height)
            .map_err(|err| WalletDbError::ZcoinClientInitError(ZcoinClientInitError::ZcashDBError(err)))?,
    };

    checkpoint_block_from_height(sync_height, rpc).await?
},
```

Reply:
done. thanks